clustering_code/edchan-kmeans.qmd

```diff
@@ -0,0 +1,946 @@
+* Use expressive names for variables so you or the reader knows what they are. 
+* Use comments to explain your code -- e.g. 
+  ```python
+  # This is a comment, it starts with a hash sign
+  ```
+
+## Label and describe your plots
+
+![Source: https://xkcd.com/833](https://imgs.xkcd.com/comics/convincing.png)
+
+Make sure each plot has:
+  * A title that explains where and when the data are from
+  * x- and y- axis labels with **units** where appropriate
+  * A legend where appropriate
+
+
+## Icons: how to use this notebook
+We use the following icons to let you know when you need to change something to complete the challenge:
+  * &#128187; means you need to write or edit some code.
+  
+  * &#128214;  indicates recommended reading
+  
+  * &#9998; marks written responses to questions
+  
+  * &#127798; is an optional extra challenge
+  
+
+---
+
+# Land cover classification at the Mississppi Delta
+
+In this notebook, you will use a k-means **unsupervised** clustering algorithm to group pixels by similar spectral signatures. k-means is an **exploratory** method for finding patterns in data. Because it is unsupervised, you don't need any training data for the model. You also can't tell how well it "performs" because the clusters will not correspond to any particular land cover class. However, we expect at least some of the clusters to be identifiable.
+
+You will use the [harmonized Sentinal/Landsat multispectral dataset](https://lpdaac.usgs.gov/documents/1698/HLS_User_Guide_V2.pdf). You can access the data with an [Earthdata account](https://www.earthdata.nasa.gov/learn/get-started) by installing the [`earthaccess` library from NSIDC](https://github.com/nsidc/earthaccess):
+
+```bash
+conda install -c conda-forge -y earthaccess
+```
+
+## STEP 1: SET UP
+
+YOUR TASK:
+  1. Import all libraries you will need for this analysis
+  2. Configure GDAL parameters to help avoid connection errors:
+     ```python
+     os.environ["GDAL_HTTP_MAX_RETRY"] = "5"
+     os.environ["GDAL_HTTP_RETRY_DELAY"] = "1"
+     ```
+
+```{python}
+#| deletable: false
+#| nbgrader: {cell_type: code, checksum: be894650820044a2987f2d17b918a09f, grade: false, grade_id: import-ans, locked: false, schema_version: 3, solution: true, task: false}
+#| tags: []
+import glob
+import itertools
+import json
+import logging
+import os
+import shutil
+import warnings
+from datetime import datetime
+
+import cartopy.crs as ccrs
+import earthpy as et
+import earthpy.plot as ep
+import earthaccess
+import geopandas as gpd
+import geoviews as gv
+import holoviews as hv
+import hvplot
+import hvplot.pandas
+import hvplot.xarray
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import pystac_client
+import rasterio as rio
+import requests
+import rioxarray as rxr
+import rioxarray.merge as rxrmerge
+import xarray as xr
+from bokeh.util import logconfig
+from dotenv import load_dotenv
+from shapely.geometry import Polygon
```

Comment:
I'm not completely sure of this, but I think the 'from' single-function imports should be separated and indented, to be in line with PEP-8 style guidelines.